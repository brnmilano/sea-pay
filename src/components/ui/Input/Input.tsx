import type { KeyboardEvent } from "react";
import type { InputProps } from "./Input.types";
import { useState, useCallback, useMemo, useId } from "react";
import { Lock, LockKeyhole } from "lucide-react";
import styles from "./input.module.scss";

/**
 * @description Componente de Input. Possui visibilidade de senha, acessibilidade,
 * estilos personalizáveis e utiliza os hooks useState, useCallback e useMemo.
 * - Utiliza o useCallback para evitar recriações desnecessárias de funções.
 * - Utiliza o useMemo para evitar recriações desnecessárias.
 * - Utiliza o useId para gerar IDs únicos para acessibilidade.
 *
 * @param param0 [InputProps]
 * @param param0.icon Ícone a ser exibido no input
 * @param param0.iconPosition Posição do ícone no input
 * @param param0.containerClassName Classes CSS personalizadas para o wrapper externo do input
 * @param param0.className Classes CSS adicionais para o input
 * @param param0.type Tipo do input sendo text ou password (o padrão é "text")
 * @param rest Todas as outras propriedades do input HTML (placeholder, disabled, onChange, etc.)
 * @returns
 */
export function Input({
  icon,
  iconPosition = "right",
  containerClassName,
  className,
  type = "text",
  ...rest
}: InputProps) {
  const [showPassword, setShowPassword] = useState<boolean>(false);
  const isPasswordType = type === "password";
  const inputId = useId();

  const inputType = isPasswordType
    ? showPassword
      ? "text"
      : "password"
    : type;

  const togglePasswordVisibility = useCallback(() => {
    setShowPassword((prev) => !prev);
  }, []);

  const handleKeyDown = useCallback(
    (event: KeyboardEvent<HTMLSpanElement>) => {
      if (event.key === "Enter" || event.key === " ") {
        event.preventDefault();
        togglePasswordVisibility();
      }
    },
    [togglePasswordVisibility],
  );

  const containerClasses = useMemo(
    () => `${styles.container} ${containerClassName}`.trim(),
    [containerClassName],
  );

  const inputClasses = useMemo(
    () => `${styles.input} ${className}`.trim(),
    [className],
  );

  const iconClasses = useMemo(() => `${styles.icon} ${styles.clickable}`, []);

  const PadlockIcon = useMemo(
    () => (showPassword ? <LockKeyhole size={20} /> : <Lock size={20} />),
    [showPassword],
  );

  const accessibilityProps = useMemo(
    () => ({
      "aria-label": showPassword ? "Ocultar senha" : "Mostrar senha",
      "aria-pressed": showPassword,
      "aria-controls": inputId,
      title: showPassword ? "Ocultar senha" : "Mostrar senha",
    }),
    [showPassword, inputId],
  );

  return (
    <div className={containerClasses}>
      <div className={styles.content}>
        <input
          id={inputId}
          className={inputClasses}
          type={inputType}
          {...rest}
        />

        {isPasswordType && (
          <span
            className={iconClasses}
            onClick={togglePasswordVisibility}
            role="button"
            tabIndex={0}
            onKeyDown={handleKeyDown}
            {...accessibilityProps}
          >
            {PadlockIcon}
          </span>
        )}

        {icon && iconPosition === "right" && !isPasswordType && (
          <span className={styles.icon}>{icon}</span>
        )}
      </div>
    </div>
  );
}
