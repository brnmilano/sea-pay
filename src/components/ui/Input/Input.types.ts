/* eslint-disable @typescript-eslint/no-explicit-any */
import type { InputHTMLAttributes, ReactNode } from "react";
import { FieldErrors, Path } from "react-hook-form";

/**
 * Posições possíveis para o ícone no input
 */
export type IconPosition = "right";

/**
 * Props para o componente Input
 * Extende as propriedades nativas do elemento input HTML
 */
export interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  /** Ícone a ser exibido no input */
  icon?: ReactNode;

  /** Posição do ícone no input */
  iconPosition?: IconPosition;

  /** Classes CSS adicionais para o container */
  containerClassName?: string;

  /** Label do input */
  label?: string;

  /**
   * Matriz de erros gerada pela biblioteca React Hook Form. É usado para renderizar o erro de entrada, se houver algum.
   */
  errors: FieldErrors<any>;
  /**
   * Propriedade utilizada para identificar a entrada do valor do input, também é utilizada no atributo "nome".
   * Se não fornecido, será usado o valor da prop 'name' do input.
   */
  registerField?: Path<any>;

  /** Referência para o elemento input */
  ref?: React.Ref<HTMLInputElement>;
}

/**
 * Props para acessibilidade do botão de toggle de senha
 */
export interface PasswordToggleAccessibilityProps {
  "aria-label": string;
  "aria-pressed": boolean;
  "aria-controls": string;
  title: string;
}
