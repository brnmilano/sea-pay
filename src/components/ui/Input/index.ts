// Barrel export para o componente Input
export { Input } from "./Input";
export type { InputProps } from "./Input.types";
