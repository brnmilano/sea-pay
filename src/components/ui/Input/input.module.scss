.container {
  display: flex;
  flex-direction: column;
  width: 100%;
  position: relative;
}

.label {
  margin-bottom: 0.4375rem;
}

.content {
  display: flex;
  align-items: center;
  border-radius: 0.375rem;
  border: 1px solid var(--text-light);
  background: var(--background);
  position: relative;
  width: 100%;
  font-weight: 400;
  min-height: 2.1875rem;
  transition:
    border-color 0.2s ease-in-out,
    box-shadow 0.2s ease-in-out;

  &:focus-within {
    border-color: var(--text-light);
    box-shadow: 0 0 0 0.5px var(--text-light);
  }

  .input {
    flex: 1;
    padding: 0.75rem 0.625rem;
    font-family: var(--font-family);
    font-size: 0.75rem;
    border: none;
    outline: none;
    background: transparent;
    color: var(--text-default);
  }

  .icon {
    display: flex;
    align-items: center;
    justify-content: center;
    color: var(--text-light);

    svg {
      width: 1.25rem;
      height: 1.25rem;
    }

    &::placeholder {
      color: var(--text-light);
    }

    &.clickable {
      cursor: pointer;
      margin-right: 1rem;
      transition: color 0.2s ease-in-out;

      &:hover {
        opacity: 0.8;
      }

      &:focus-visible {
        outline: 0.125rem solid var(--text-light);
        outline-offset: 0.125rem;
        border-radius: 0.125rem;
      }
    }
  }
}

.errorMessage {
  color: var(--semantic-red);
  margin-top: -10px;
  visibility: hidden;
  opacity: 0;
  position: absolute;
  bottom: -30px;

  p {
    font-size: 0.75rem !important;
  }
}

.showingErrorMessage {
  visibility: visible;
  opacity: 1;
  margin-top: 10px;
}
