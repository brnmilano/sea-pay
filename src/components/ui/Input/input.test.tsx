import { render, screen, fireEvent } from "@testing-library/react";
import { Input } from "./Input";

describe("Input Component", () => {
  test("should match snapshot.", () => {
    const { container } = render(
      <Input
        label="Snapshot Input"
        placeholder="Snapshot Input"
        aria-description="Snapshot Input"
        registerField="snapshot"
        errors={{}}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test("should check if the passed ref points to the correct DOM element.", () => {
    const ref = { current: null };

    render(
      <Input
        label="Ref Input"
        placeholder="Ref Input"
        aria-description="Ref Input"
        registerField="ref"
        errors={{}}
        ref={ref}
      />,
    );

    const input = screen.getByPlaceholderText("Ref Input");

    expect(input).toBeInTheDocument();
    expect(ref.current).toBe(input);
  });

  test("should render the input with correct label, placeholder and aria-description.", () => {
    render(
      <Input
        label="Test Input"
        placeholder="Enter text"
        aria-description="Test Input"
        registerField="test"
        errors={{}}
      />,
    );

    const input = screen.getByPlaceholderText("Enter text");
    const label = screen.getByText("Test Input");
    const ariaDescription = screen.getByLabelText("Test Input");

    expect(input).toBeInTheDocument();
    expect(label).toBeInTheDocument();
    expect(ariaDescription).toHaveAttribute("aria-description", "Test Input");
    expect(label).toHaveAttribute("for", input.id);
  });

  test("should render the input with the errors and registerField props.", () => {
    const errors = {
      test: {
        type: "required",
        message: "Required field",
      },
    };

    render(
      <Input
        label="Test Input"
        placeholder="Enter text"
        aria-description="Test Input"
        registerField="test"
        errors={errors}
      />,
    );

    const errorMessage = screen.getByText("Required field");
    const errorContainer = errorMessage.parentElement;

    expect(errorMessage).toBeInTheDocument();
    expect(errorContainer).toHaveClass("errorMessage");
    expect(errorContainer).toHaveClass("showingErrorMessage");
  });

  test("should render the correct defaultValue.", () => {
    render(
      <Input
        label="Test Input"
        placeholder="Enter text"
        aria-description="Test Input"
        registerField="test"
        errors={{}}
        defaultValue="Default Value"
      />,
    );

    const input = screen.getByPlaceholderText("Enter text") as HTMLInputElement;

    expect(input).toBeInTheDocument();
    expect(input).toHaveValue("Default Value");
  });

  test("should check keyboard focus (tab) and visible focus indicator.", () => {
    render(
      <Input
        label="Test Input"
        placeholder="Enter text"
        aria-description="Test Input"
        registerField="test"
        errors={{}}
      />,
    );

    const input = screen.getByPlaceholderText("Enter text");

    expect(input).toBeInTheDocument();
    expect(input).not.toHaveFocus();
    expect(document.body).toHaveFocus();

    input.focus();
    expect(input).toHaveFocus();
    expect(document.body).not.toHaveFocus();
  });

  test("should check the integration with react-hook-form by using the registerField prop.", () => {
    const handleChange = jest.fn();

    render(
      <Input
        label="Test Input"
        placeholder="Enter text"
        aria-description="Test Input"
        registerField="test"
        errors={{}}
        onChange={handleChange}
      />,
    );

    const input = screen.getByPlaceholderText("Enter text") as HTMLInputElement;

    expect(input).toBeInTheDocument();

    input.focus();
    expect(input).toHaveFocus();

    fireEvent.change(input, { target: { value: "RHF Test" } });
    expect(handleChange).toHaveBeenCalledTimes(1);
    expect(input).toHaveValue("RHF Test");
  });

  test("should simulate typing behavior and ensure that onChange is called with the expected value.", () => {
    const handleChange = jest.fn();

    render(
      <Input
        label="Test Input"
        placeholder="Enter text"
        aria-description="Test Input"
        registerField="test"
        errors={{}}
        onChange={handleChange}
      />,
    );

    const input = screen.getByPlaceholderText("Enter text") as HTMLInputElement;

    expect(input).toBeInTheDocument();

    input.focus();
    expect(input).toHaveFocus();

    const text = "000.000.000-00";

    for (let i = 0; i < text.length; i++) {
      const newValue = text.slice(0, i + 1);

      fireEvent.change(input, { target: { value: newValue } });

      expect(handleChange).toHaveBeenLastCalledWith(
        expect.objectContaining({
          target: expect.objectContaining({ value: newValue }),
        }),
      );
    }

    expect(input).toHaveValue("000.000.000-00");
  });
});
