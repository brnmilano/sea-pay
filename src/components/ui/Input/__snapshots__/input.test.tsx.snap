// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Input Component should match snapshot. 1`] = `
<div>
  <div
    class="container undefined"
  >
    <label
      class="label"
      for="«r0»"
    >
      Snapshot Input
    </label>
    <div
      class="content"
    >
      <input
        aria-description="Snapshot Input"
        class="input undefined"
        id="«r0»"
        placeholder="Snapshot Input"
        type="text"
      />
    </div>
    <div
      class="errorMessage undefined"
    >
      <p />
    </div>
  </div>
</div>
`;
